File: LeetCode/985.bag-of-tokens/bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int P) {
        if (tokens == null || tokens.length == 0)   return 0;
        Arrays.sort(tokens);
        int length = tokens.length;
        int point = 0;
        int start = 0, end = length - 1;
        for (;;){
            if (start > end || P < tokens[start])   break;
            while (start <= end && P >= tokens[start]){
                point++;
                P -= tokens[start++];
            }
            while (point > 0 && start < end && P < tokens[start]){
                point--;
                P += tokens[end--];
            }
        }
        return point;
    }
}

File: LeetCode/0948.bag-of-tokens/bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int P) {
        int point = 0;
        int sum = 0;
        int size = tokens.size();
        for (int i = 0; i < size; i++) sum += tokens[i];
        int start = 0, end = size - 1;
        sort(tokens.begin(), tokens.end());
        for (;;){
            if (start > end || P < tokens[start])    break;
            while (start <= end && P >= tokens[start]){
                point++;
                P -= tokens[start];
                start++;
            }
            while (point > 0 && P < tokens[start] && start < end){
                point--;
                P += tokens[end];
                end--;
            }
        }
        return point;
    }
};